File: javascript/tests/discovery.test.js
/**
 * Tests for discovery document handling.
 */

import { test, describe } from 'node:test';
import assert from 'node:assert';
import { EntityType, AgentStatus } from '../src/types.js';
import {
    buildDiscoveryDocument,
    validateDiscoveryDocument,
    findKeyByKid,
    findAgentById,
} from '../src/discovery.js';

function makeTestDoc() {
    return buildDiscoveryDocument(
        'example.com',
        EntityType.MAKER,
        [{
            kid: 'example-2026-01',
            kty: 'EC',
            crv: 'P-256',
            x: 'test-x',
            y: 'test-y',
            use: 'sig',
            key_ops: ['verify'],
        }],
        [{
            agent_id: 'urn:agentpin:example.com:agent',
            name: 'Test Agent',
            capabilities: ['read:*'],
            status: AgentStatus.ACTIVE,
        }],
        2,
        '2026-01-15T00:00:00Z'
    );
}

describe('validateDiscoveryDocument', () => {
    test('valid document passes', () => {
        const doc = makeTestDoc();
        assert.doesNotThrow(() => validateDiscoveryDocument(doc, 'example.com'));
    });

    test('domain mismatch rejected', () => {
        const doc = makeTestDoc();
        assert.throws(
            () => validateDiscoveryDocument(doc, 'other.com'),
            /does not match/
        );
    });
});

describe('findKeyByKid', () => {
    test('finds existing key', () => {
        const doc = makeTestDoc();
        assert.ok(findKeyByKid(doc, 'example-2026-01'));
    });

    test('returns null for missing key', () => {
        const doc = makeTestDoc();
        assert.strictEqual(findKeyByKid(doc, 'nonexistent'), null);
    });
});

describe('findAgentById', () => {
    test('finds existing agent', () => {
        const doc = makeTestDoc();
        assert.ok(findAgentById(doc, 'urn:agentpin:example.com:agent'));
    });

    test('returns null for missing agent', () => {
        const doc = makeTestDoc();
        assert.strictEqual(findAgentById(doc, 'urn:agentpin:example.com:other'), null);
    });
});

File: go/pkg/discovery/discovery_test.go
package discovery

import (
	"testing"

	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
)

func makeDoc() types.DiscoveryDocument {
	return BuildDiscoveryDocument(
		"example.com",
		types.EntityMaker,
		[]types.JWK{{Kid: "example-2026-01", Kty: "EC", Crv: "P-256", X: "x", Y: "y", Use: "sig"}},
		[]types.AgentDeclaration{
			{
				AgentID:      "urn:agentpin:example.com:agent",
				Name:         "Test Agent",
				Capabilities: []types.Capability{"read:*"},
				Status:       types.AgentActive,
			},
		},
		2,
		"2026-01-15T00:00:00Z",
	)
}

func TestValidateDiscoveryDocument(t *testing.T) {
	d := makeDoc()
	if err := ValidateDiscoveryDocument(&d, "example.com"); err != nil {
		t.Fatal(err)
	}
}

func TestValidateDomainMismatch(t *testing.T) {
	d := makeDoc()
	err := ValidateDiscoveryDocument(&d, "other.com")
	if err == nil {
		t.Fatal("expected mismatch error")
	}
	ve, ok := types.AsVerificationError(err)
	if !ok || ve.Code != types.ErrDomainMismatch {
		t.Fatalf("expected DomainMismatch, got %v", err)
	}
}

func TestFindKeyByKid(t *testing.T) {
	d := makeDoc()
	if FindKeyByKid(&d, "example-2026-01") == nil {
		t.Fatal("hit expected")
	}
	if FindKeyByKid(&d, "missing") != nil {
		t.Fatal("miss expected")
	}
}

func TestFindAgentByID(t *testing.T) {
	d := makeDoc()
	if FindAgentByID(&d, "urn:agentpin:example.com:agent") == nil {
		t.Fatal("hit expected")
	}
	if FindAgentByID(&d, "urn:agentpin:example.com:other") != nil {
		t.Fatal("miss expected")
	}
}

func TestBuildDiscoveryDocumentRevocationEndpoint(t *testing.T) {
	d := makeDoc()
	want := "https://example.com/.well-known/agent-identity-revocations.json"
	if d.RevocationEndpoint != want {
		t.Fatalf("RevocationEndpoint = %s, want %s", d.RevocationEndpoint, want)
	}
}

func TestRejectMaxDelegationDepthOver3(t *testing.T) {
	d := makeDoc()
	d.MaxDelegationDepth = 4
	if err := ValidateDiscoveryDocument(&d, "example.com"); err == nil {
		t.Fatal("max_delegation_depth > 3 should be rejected")
	}
}

func TestRejectEmptyKeys(t *testing.T) {
	d := makeDoc()
	d.PublicKeys = nil
	if err := ValidateDiscoveryDocument(&d, "example.com"); err == nil {
		t.Fatal("empty keys should be rejected")
	}
}
